integrated_channels/integrated_channel/models.py

```diff
@@ -60,5 +78,141 @@ class EnterpriseCustomerPluginConfiguration(TimeStampedModel):
+        """
+        raise NotImplementedError(_('Implemented in concrete subclass.'))
+
+    def transmit_learner_data(self, api_username):
+        """
+        Collect and transmit learner data for the ``EnterpriseCustomer``.
+
+        Arguments:
+
+        * ``api_username``: Username of a user authorized to fetch grades from the LMS API.
+        """
+        transmitter = self.get_learner_data_transmitter()
+        for learner_data in self.collect_learner_data(api_username):
+            transmitter.transmit(learner_data)
+
+    def collect_learner_data(self, api_username):
+        """
+        Collect learner data for the ``EnterpriseCustomer`` where data sharing consent is granted.
+
+        Arguments:
+
+        * ``api_username``: Username of a user authorized to fetch grades from the LMS API.
+
+        Yields a learner data object for each enrollment, containing:
+
+        * ``enterprise_enrollment``: ``EnterpriseCourseEnrollment`` object.
+        * ``certificate``: ``GeneratedCertificate`` for the user+course; None if not found.
+          "Course completion" occurs when course certificates are issued.
+        """
+
+        if CourseKey is None or GeneratedCertificate is None:
+            raise NotConnectedToOpenEdX(_('This package must be installed in an OpenEdX environment.'))
+
+        # Fetch course details from the Course API
+        course_api = CourseApiClient()
+        course_details = None
+
+        # Fetch Grade details from the Grades API
+        grades_api = GradesApiClient(api_username)
+
+        # Fetch the consenting enrollment data, including the enterprise_customer_user.
+        # Order by the course_id, to avoid fetching course API data more than we have to.
+        enrollment_queryset = EnterpriseCourseEnrollment.objects.select_related(
+            'enterprise_customer_user'
+        ).filter(
+            enterprise_customer_user__enterprise_customer=self.enterprise_customer,
+        ).order_by('course_id')
+        for enterprise_enrollment in enrollment_queryset:
+
+            # Omit any enrollments where consent has not been granted
+            if not enterprise_enrollment.consent_available():
+                continue
+
+            course_id = enterprise_enrollment.course_id
+            completed_date = None
+            grade = None
+
+            # Fetch the user+course certificate, if found
+            try:
+                # If a certificate is found, then we use that to determine course completion.
+                certificate = GeneratedCertificate.eligible_certificates.get(
+                    user__id=enterprise_enrollment.enterprise_customer_user.user_id,
+                    course_id=CourseKey.from_string(course_id),
+                )
+                completed_date = certificate.created_date
+                # TODO: change to Pass/Fail/In Progress?
+                grade = certificate.grade
+
+            except GeneratedCertificate.DoesNotExist:
```

Comment:
@pomegranited 40 lines long _exception handling_ - probably needs a method (and probably more than one)